File: 19-Iterator/src/go/playlist.go
package main

// Playlist is the Aggregate — it stores songs and vends iterators so clients
// never touch the internal slice.
type Playlist struct {
	songs []Song
}

func (p *Playlist) AddSong(song Song) {
	p.songs = append(p.songs, song)
}

func (p *Playlist) Iterator() Iterator {
	return &ForwardIterator{songs: p.songs, index: 0}
}

func (p *Playlist) ReverseIterator() Iterator {
	return &ReverseIterator{songs: p.songs, index: len(p.songs) - 1}
}

File: 19-Iterator/src/rust/playlist.rs
use crate::song::Song;
use crate::forward_iterator::ForwardIterator;
use crate::reverse_iterator::ReverseIterator;

/// Aggregate — stores songs and vends iterators so clients never touch
/// the internal Vec.
pub struct Playlist {
    songs: Vec<Song>,
}

impl Playlist {
    pub fn new() -> Self {
        Playlist { songs: Vec::new() }
    }

    pub fn add_song(&mut self, song: Song) {
        self.songs.push(song);
    }

    pub fn iter(&self) -> ForwardIterator {
        ForwardIterator::new(&self.songs)
    }

    pub fn reverse_iter(&self) -> ReverseIterator {
        ReverseIterator::new(&self.songs)
    }
}
